src/rcc.rs

```diff
@@ -214,91 +231,130 @@ impl CFGR {
             return (pllsrcclk, None);
         }
 
         let pllmul = cmp::min(cmp::max(pllmul, 2), 16);
-        let sysclk = pllmul * pllsrcclk;
+        let sysclk = pllsrcclk * pllmul;
         assert!(sysclk <= 72_000_000);
 
-        let pllmul_bits = pllmul as u8 - 2;
-        (sysclk, Some((pllmul_bits, pllsrc)))
+        // NOTE From<u8> for PLLMUL_A is not implemented
+        // https://docs.rs/stm32f3/0.9.0/src/stm32f3/stm32f303/rcc/cfgr.rs.html#629-662
+        // let pllmul_mul: cfgr::PLLMUL_A = (pllmul as u8 - 2).into();
+        let pllmul_mul = match pllmul as u8 {
+            2 => cfgr::PLLMUL_A::MUL2,
+            3 => cfgr::PLLMUL_A::MUL3,
+            4 => cfgr::PLLMUL_A::MUL4,
+            5 => cfgr::PLLMUL_A::MUL5,
+            6 => cfgr::PLLMUL_A::MUL6,
+            7 => cfgr::PLLMUL_A::MUL7,
+            8 => cfgr::PLLMUL_A::MUL8,
+            9 => cfgr::PLLMUL_A::MUL9,
+            10 => cfgr::PLLMUL_A::MUL10,
+            11 => cfgr::PLLMUL_A::MUL11,
+            12 => cfgr::PLLMUL_A::MUL12,
+            13 => cfgr::PLLMUL_A::MUL13,
+            14 => cfgr::PLLMUL_A::MUL14,
+            15 => cfgr::PLLMUL_A::MUL15,
+            16 => cfgr::PLLMUL_A::MUL16,
+            17 => cfgr::PLLMUL_A::MUL16X,
+            _ => unreachable!(),
+        };
+        (sysclk, Some((pllmul_mul, pllsrc)))
     }
 
     /// Freezes the clock configuration, making it effective
     pub fn freeze(self, acr: &mut ACR) -> Clocks {
         let (sysclk, pll_options) = self.calc_sysclk();
 
-        let hpre_bits = self
+        let hpre_div = self
             .hclk
             .map(|hclk| match sysclk / hclk {
                 0 => unreachable!(),
-                1 => 0b0111,
-                2 => 0b1000,
-                3..=5 => 0b1001,
-                6..=11 => 0b1010,
-                12..=39 => 0b1011,
-                40..=95 => 0b1100,
-                96..=191 => 0b1101,
-                192..=383 => 0b1110,
-                _ => 0b1111,
+                1 => cfgr::HPRE_A::DIV1,
+                2 => cfgr::HPRE_A::DIV2,
+                3..=5 => cfgr::HPRE_A::DIV4,
+                6..=11 => cfgr::HPRE_A::DIV8,
+                12..=39 => cfgr::HPRE_A::DIV16,
+                40..=95 => cfgr::HPRE_A::DIV64,
+                96..=191 => cfgr::HPRE_A::DIV128,
+                192..=383 => cfgr::HPRE_A::DIV256,
+                _ => cfgr::HPRE_A::DIV512,
             })
-            .unwrap_or(0b0111);
+            .unwrap_or(cfgr::HPRE_A::DIV1);
+
+        // Get the shift factor by subtracting the offset
+        //
+        // `DIV2 - 1` is used for offset, as this is guaranteed to have only one value
+        // as opposed to `DIV1` which can be 0..=7
+        let hpre_div_factor = u8::from(hpre_div)
+            .checked_sub(u8::from(cfgr::HPRE_A::DIV2) - 1)
+            .unwrap_or_default();
 
-        let hclk = sysclk / (1 << (hpre_bits - 0b0111));
+        let hclk: u32 = sysclk / 1 << hpre_div_factor;
```

Comment:
I think there's a (pre-existing) bug in this line: Notice how `DIV32` does not exist (I also confirmed this in the reference manual). If I understand things correctly, that means that for `DIV64` and up the result here will be incorrect.

Reply:
Good catch. Yes you are right. I'll try to fix this

Reply:
I've added a special case. If you want, you could take another look @TimNN

Reply:
BTW the special case was missing before my change as well, so this should fix a bug which was there for some time.